### ISSUE TITLE
3 features: broadcast more events, add strategy name to the broadcast event.data, add networkTimeoutSeconds to StaleWhileRevalidate

### ISSUE THREAD

**Issue by johnrobertcobbold** [GoogleChrome/workbox#2010]
**Library Affected**:
workbox-strategies

**Browser & Platform**:
"all browsers".

**Issue or Feature Request Description**:
Some of our pages use the StaleWhileRevalidate strategy. There does not seem to be a way to specifically request to bypass the cache if we *know* that the page has changed. In those cases, displaying the cached page for a few seconds before refreshing it with the newly fetched data (using broadcastUpdate) confuses the user who is also not expecting to see old data.

**Comment by jeffposnick** [comment]
The simplest approach would probably be to purge all of the previously cached responses if you have a way of knowing that they're no longer valid.

For instance, if you use `cacheName: 'page-cache'` along with your stale-while-revalidate strategy, once you detect that they're no longer fresh, you could use `caches.delete('page-cache')` to clear everything out. You could run that code from either inside of the service worker, or from the context of your client web app, since the Cache Storage API is exposed on the `window` object.

Alternatively, you could write a custom plugin that used the [`cachedResponseWillBeUsed` lifecycle method](https://developers.google.com/web/tools/workbox/guides/using-plugins#custom_plugins) to determine whether a cached response is fresh or not, and return `null` if you have determined that it's stale. But I think it would be cleaner to just purge everything at the point in time that you determine the previous responses are no longer fresh.

Hopefully that gives you something to go on. I'm going to close this issue, but if there's a reason why those suggestions don't work, we can revisit.

**Comment by goyavo** [comment]
Hello Jeff,

Indeed, when we know that the page has changed, we purge the cache using the Cache Storage API *before* fetching that url, so that StaleWhileRevalidate has no choice but to fetch from the network and the user does not first see an old cached version.
What happens though if at that moment the device goes offline (I'm in Cambodia where the network is often erratic)? Nothing loads but now things are worse, we removed the cached version so there is nothing left to show (－‸ლ)  !

After thinking about this, I believe that the following could be the answer:
> allow strategies to use broadcastUpdate to broadcast more events:

1a.  StaleWhileRevalidate: "I found a cached version" (type: 'CACHE_FOUND')
> page displays cached version (or perhaps chooses not to show anything yet)

2a: or StaleWhileRevalidate: "I did find any cached version"  (type: 'CACHE_NOT_FOUND')

2. StaleWhileRevalidate: "Now, I am trying to fetch an updated version"  (type: 'FETCHING_FROM_NETWORK')
> page displays an "updating" banner

3a StaleWhileRevalidate: "I was able to fetch from the network"  (type: 'FETCHING_SUCCESS')
> page stops displaying the "updating" banner

3b. or StaleWhileRevalidate: "darn, the network is down!" (type: 'FETCHING_ERROR')
> page displays some message i.e "network error, this is a cached version" (and displays the cached version if it was not already displayed in step 1)
-----------------

if (3a):

4a. StaleWhileRevalidate: "... and I found a cool updated version"  (type: 'CACHE_UPDATED')
> page updates content (this one is already possible with broadcastUpdate)

4b: StaleWhileRevalidate does not find anything new. Nothing is broadcasted.

In other words, the feature request is to allow StaleWhileRevalidate, NetworkFirst and CacheFirst to use the broadcastUpdate.Plugin to broadcast more events. Example:
"CACHE_FOUND"
"CACHE_NOT_FOUND"
'FETCHING_FROM_NETWORK"
"FETCHING_SUCCESS"
"FETCHING_ERROR"
"CACHE_UPDATED"

**Comment by goyavo** [comment]
@jeffposnick let me know if it makes sense to open 3 new tickets for this?

**Comment by philipwalton** [comment]
Hey @goyavo thanks for the detailed suggestions. I think your use case is totally valid; though, rather than adding a (potentially endless) list of new update events to broadcast, I think we can address this issue in a more generic way by making it easier to customize what data gets sent in the broadcast update plugin as well as making it easier (in general) for code in the SW to talk to the window.

Here are two existing issues that I believe will help address your use case:

- Make it easier to customize the broadcast update payload. https://github.com/GoogleChrome/workbox/issues/1878
- Make it easier to generically send messages between the SW to the window. https://github.com/GoogleChrome/workbox/issues/1848

What do you think?

**Comment by goyavo** [comment]
Hello @philipwalton, 

#1878, indeed sounds like something similar, on the condition that `workbox-broadcast-cache-update` could start broadcasting events from the very beginning, such as "1a: StaleWhileRevalidate found a cached version" (see my previous timeline), and not only once the `newResponse` from your example was fetched. Would it really be a endless list of update events? My previous post had 5 new events out of a total of 6 (since CACHE_UPDATED already exists).

**Comment by tomayac** [comment]
Hi there,

Workbox is [moving to a new engineering team within Google](https://github.com/GoogleChrome/workbox?tab=readme-ov-file#maintenance-update:~:text=Maintenance%20update,owners%20of%20Workbox.). As part of this move, we're declaring a partial bug bankruptcy to allow the new team to start fresh. We realize this isn't optimal, but realistically, this is the only way we see it working. For transparency, here're the criteria we applied:

- For PRs, we closed [everything that doesn't target the current `v7` branch](https://github.com/GoogleChrome/workbox/pulls?q=is%3Apr+is%3Aclosed+-base%3Av7).
- For Issues, we closed [everything that was created before 2024](https://github.com/GoogleChrome/workbox/issues?q=is%3Aissue+is%3Aclosed+created%3A%3C2024-01-01+). 

Thanks, and we hope for your understanding!
The Workbox team
